README
NAME
    Object::FromData - Throw arrayrefs or hashrefs at this to get an object

VERSION
    Version 0.01

SYNOPSIS
        use Object::FromData;

        my $object = Object::FromData->new(
            {   ref => {
                    foo      => 'bar',
                    this     => 'that',
                    _private => 'no method will be created',
                    colors   => [qw/red blue green/],
                    numbers  => {
                        un    => 1,
                        deux  => 2,
                        trois => 3,
                    },
                },
            }
        );

        say $object->is_hashref;     # 1
        say $object->is_arrayref;    # 0

        say $object->foo;            # bar
        say $object->this;           # that
        say $object->_private;       # BOOM! No such method

        say $object->numbers->is_hashref;     # 0
        say $object->numbers->is_arrayref;    # 1
        say $object->numbers->deux;           # 2

        my $colors = $object->colors;
        while ( $colors->has_more ) {
            say $colors->next;    # red, blue, green on successive lines
        }

  "new"
        my $object = Object::FromData->new({ ref => $ref });

    Takes a hashref. The key of "ref" must be a hashref or arrayref. Returns
    an object. The object represents a hash or an arrayref as created by
    "Object::FromData::Hash" or "Object::FromData::Array".

    The primary purpose of this module was to be able to take arbitrary JSON
    from a client and create an object I could safely embed in another
    object.

AUTHOR
    Curtis "Ovid" Poe, "<ovid at cpan.org>"

BUGS
    Please report any bugs or feature requests to "bug-object-fromdata at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Object-FromData>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

    We've tried to keep the methods minimal, but because we're inheriting
    from "Object::FromData::Array" and "Object::FromData::Hash", the latter
    might have keys which override the main keys. If that happens, call the
    methods as class methods, passing in the object as an argument:

        my %hash = (
            keys => [qw/foo bar baz/],
        );
        my $object = Object::FromData->new({ ref => \%hash });

        my @keys = $hash->keys; # returns an Object::FromData::Array instance of foo, bar, and baz
        my @keys = Object::FromData::Hash->keys($hash); # returns 'keys'

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Object::FromData

    You can also look for information at:

    *   RT: CPAN's request tracker (report bugs here)

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Object-FromData>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Object-FromData>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Object-FromData>

    *   Search CPAN

        <http://search.cpan.org/dist/Object-FromData/>

ACKNOWLEDGEMENTS
LICENSE AND COPYRIGHT
    Copyright 2016 Curtis "Ovid" Poe.

    This program is free software; you can redistribute it and/or modify it
    under the terms of the the Artistic License (2.0). You may obtain a copy
    of the full license at:

    <http://www.perlfoundation.org/artistic_license_2_0>

    Any use, modification, and distribution of the Standard or Modified
    Versions is governed by this Artistic License. By using, modifying or
    distributing the Package, you accept this license. Do not use, modify,
    or distribute the Package, if you do not accept this license.

    If your Modified Version has been derived from a Modified Version made
    by someone other than you, you are nevertheless required to ensure that
    your Modified Version complies with the requirements of this license.

    This license does not grant you the right to use any trademark, service
    mark, tradename, or logo of the Copyright Holder.

    This license includes the non-exclusive, worldwide, free-of-charge
    patent license to make, have made, use, offer to sell, sell, import and
    otherwise transfer the Package with respect to any patent claims
    licensable by the Copyright Holder that are necessarily infringed by the
    Package. If you institute patent litigation (including a cross-claim or
    counterclaim) against any party alleging that the Package constitutes
    direct or contributory patent infringement, then this Artistic License
    to you shall terminate on the date that such litigation is filed.

    Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
    AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
    THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
    PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
    YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
    CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
    EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Object-FromData

0.03    2016-09-01
        - Make it work on older Perls. (sort on unknown subroutine bug)
        - Tighten up some internals.

0.02    2016-08-31 19:16:35
        - Get rid of each() iterator on hash
        - Fix debugging statement in t/hash.t whose warning caused a faillure on
          one system.

0.01    2016-08-31
        - Throw references at this to get objects



lib/Object/FromData/Hash.pm
package Object::FromData::Hash;

use strict;
use warnings;
use Digest::MD5 'md5_hex';    # convenience, not security

our $VERSION = '0.03';

sub _new {
    my ( $class, $arg_for ) = @_;

    my $hashref       = $arg_for->{ref};
    my $allow_private = $arg_for->{allow_private} || 0;
    my @all_keys      = sort( CORE::keys(%$hashref) );

    my $self = {
        hash          => {},
        allow_private => $allow_private,
        keys => [ $allow_private ? @all_keys : grep { !/^_/ } @all_keys ],
    };
    my $new_class = "${class}::" . md5_hex( join '-' => @{ $self->{keys} } );
    {
        no strict 'refs';
        @{"${new_class}::ISA"} = $class;
    }

    KEY: foreach my $key (@all_keys) {
        my $value = $hashref->{$key};
        if ( 'ARRAY' eq ref $value ) {
            $self->{hash}{$key}
              = Object::FromData::Array->_new( { ref => $value } );
        }
        elsif ( 'HASH' eq ref $value ) {
            $self->{hash}{$key} = $class->_new( { ref => $value } );
        }
        else {
            $self->{hash}{$key} = $value;
        }
        if ( $key =~ /^_/ ) {
            next KEY unless $allow_private;
        }
        no strict 'refs';
        *{"${new_class}::$key"} = sub { $self->{hash}{$key} };
    }
    return bless $self => $new_class;
}

sub keys {
    my $self = shift;
    $self = shift if @_;
    return keys %{ $self->{hash} };
}

sub values {
    my $self = shift;
    $self = shift if @_;
    return values %{ $self->{hash} };
}

sub is_hashref  {1}
sub is_arrayref { }

1;

__END__

=head1 NAME

Object::FromData::Hash - Create an object from a hashref.

=head1 SYNOPSIS

Don't instantiate directly using this module. Use C<Object::FromData> instead.

    my %hash = (
        name   => 'Bob',
        age    => 27,
        colors => [qw/red green blue/],
    );

    my $hashref = Object::FromData->new( { ref => \%hash } );
    my @keys    = $hashref->keys;
    my @values  = $hashref->values;
    say $hashref->name;    # Bob
    my $arrayref_object = $hashref->colors;    # Object::FromData::Array

=head2 C<is_hashref>

    if ( $hashref->is_hashref ) { ... }

Returns true.

=head2 C<is_arrayref>

    if ( $hashref->is_arrayref ) { ... }

Returns false.

=head2 C<keys>

    my @keys = $hashref->keys;

Returns a I<list> of the keys.

=head2 C<values>

    my @values = $hashref->values;

Returns a I<list> of the values.

=head1 HASH KEYS OVERRIDING BUILT IN METHODS

We've tried to keep the methods minimal, but because we're inheriting from
C<Object::FromData::Hash> and the hash you pass in might have keys which
override the main keys. If that happens, call the methods as class methods,
passing in the object as an argument:

    my %hash = (
        keys => [qw/foo bar baz/],
    );
    my $object = Object::FromData->new({ ref => \%hash });

    my @keys = $hash->keys; # returns an Object::FromData::Array instance of foo, bar, and baz
    my @keys = Object::FromData::Hash->keys($hash); # returns 'keys'


lib/Object/FromData/Array.pm
package Object::FromData::Array;

use strict;
use warnings;

our $VERSION = '0.03';

sub _new {
    my ( $class, $arg_for ) = @_;

    my $arrayref = $arg_for->{ref};
    my $self     = {
        array         => [],
        current_index => 0,
    };
    foreach my $item (@$arrayref) {
        if ( 'ARRAY' eq ref $item ) {
            push @{ $self->{array} } => $class->_new( { ref => $item } );
        }
        elsif ( 'HASH' eq ref $item ) {
            push @{ $self->{array} } =>
              Object::FromData::Hash->_new( { ref => $item } );
        }
        else {
            push @{ $self->{array} } => $item;    # accept it raw
        }
    }
    return bless $self => $class;
}

sub current_index {
    my $self = shift;
    return $self->{current_index};
}

sub num_elems {
    my $self = shift;
    return scalar @{ $self->{array} };
}

sub reset {
    my $self = shift;
    $self->{current_index} = 0;
}

sub all {
    my $self = shift;
    return @{ $self->{array} };
}

sub has_more {
    my $self     = shift;
    my $arrayref = $self->{array};
    return $self->{current_index} <= $#$arrayref;
}

sub next {
    my $self = shift;
    return unless $self->has_more;
    return $self->{array}->[ $self->{current_index}++ ];
}

sub is_hashref  { }
sub is_arrayref {1}

1;

__END__

=head1 NAME

Object::FromData::Array - Create an object from an arrayref.

=head1 SYNOPSIS

Don't instantiate directly using this module. Use C<Object::FromData> instead.

    my $arrayref = Object::FromData->new( { ref => \@array } );

    while ( $arrayref->has_more ) {
        my $value = $arrayref->next;
        ...
    }

=head2 C<has_more>

    while ( $arrayref->has_more ) {
        my $value = $arrayref->next;
        ...
    }

Boolean indicating if the array iterator is exhausted.

=head2 C<next>

    my $value = $arrayref->next;

Returns the next value in the arrayref iterator.

=head2 C<reset>

    $arrayref->reset;

Resets the arrayref iterator.

=head2 C<all>

    my @values = $arrayref->all;

Returns a list of all values. Each value I<might> be an C<Object::FromData::>
object if it's a reference.

=head2 C<num_elems>

    my $num_elems = $arrayref->num_elems;

Returns the number of elements in the array reference.

=head2 C<current_index>

    my $i = $arrayref->current_index;

Returns the current numeric index of the iterator.

=head2 C<is_hashref>

    if ( $arrayref->is_hashref ) { ... }

Returns false.

=head2 C<is_arrayref>

    if ( $arrayref->is_arrayref ) { ... }

Returns true.


lib/Object/FromData.pm
package Object::FromData;

use strict;
use warnings;
use Object::FromData::Array;
use Object::FromData::Hash;
use Carp 'confess';

=head1 NAME

Object::FromData - Throw arrayrefs or hashrefs at this to get an object

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.03';

=head1 SYNOPSIS

    use Object::FromData;

    my $object = Object::FromData->new(
        {   ref => {
                foo      => 'bar',
                this     => 'that',
                _private => 'no method will be created',
                colors   => [qw/red blue green/],
                numbers  => {
                    un    => 1,
                    deux  => 2,
                    trois => 3,
                },
            },
        }
    );

    say $object->is_hashref;     # 1
    say $object->is_arrayref;    # 0

    say $object->foo;            # bar
    say $object->this;           # that
    say $object->_private;       # BOOM! No such method

    say $object->numbers->is_hashref;     # 0
    say $object->numbers->is_arrayref;    # 1
    say $object->numbers->deux;           # 2

    my $colors = $object->colors;
    while ( $colors->has_more ) {
        say $colors->next;    # red, blue, green on successive lines
    }

=head2 C<new>

    my $object = Object::FromData->new({ ref => $ref });

Takes a hashref. The key of C<ref> must be a hashref or arrayref. Returns an
object. The object represents a hash or an arrayref as created by
C<Object::FromData::Hash> or C<Object::FromData::Array>.

The primary purpose of this module was to be able to take arbitrary JSON from
a client and create an object I could safely embed in another object.

=cut

sub new {
    my ( $class, $arg_for ) = @_;

    my $ref = $arg_for->{ref}
      or confess "no ref() supplied to Object::FromData::new";
    if ( 'ARRAY' eq ref $ref ) {
        return Object::FromData::Array->_new( { ref => $ref } );
    }
    elsif ( 'HASH' eq ref $ref ) {
        return Object::FromData::Hash->_new( { ref => $ref } );
    }
    else {
        croak(
            "Don't know how to convert '$ref' into an object. I only handle arrayrefs and hashrefs"
        );
    }
}

=head1 AUTHOR

Curtis "Ovid" Poe, C<< <ovid at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-object-fromdata at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Object-FromData>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

We've tried to keep the methods minimal, but because we're inheriting from
C<Object::FromData::Array> and C<Object::FromData::Hash>, the latter might
have keys which override the main keys. If that happens, call the methods as
class methods, passing in the object as an argument:

    my %hash = (
        keys => [qw/foo bar baz/],
    );
    my $object = Object::FromData->new({ ref => \%hash });

    my @keys = $hash->keys; # returns an Object::FromData::Array instance of foo, bar, and baz
    my @keys = Object::FromData::Hash->keys($hash); # returns 'keys'

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Object::FromData

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Object-FromData>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Object-FromData>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Object-FromData>

=item * Search CPAN

L<http://search.cpan.org/dist/Object-FromData/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2016 Curtis "Ovid" Poe.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1;    # End of Object::FromData
